### ISSUE TITLE
Memoization files Saved to User/AppData folder and not being deleted

### ISSUE THREAD

**Issue by lacan** [qupath/qupath#717]
Hi, 

I know there was such as issue described here already: https://github.com/qupath/qupath/issues/287

But I wanted to point out that with the latest 0.3.0-SNAPSHOT, the memoization file got saved in 
`C:\Users\oburri\AppData\Local\Temp\qupath-memo-15327719930616011555\biop\Image_Processing\User_Projects\Olivier\USERNAME_REDACTED\Sample images - metadata`

This mimmics the architecture I have where the actual file is stored:
`L:\biop\Image_Processing\User_Projects\Olivier\USERNAME_REDACTED\Sample images - metadata`

But in previous versions the memo files were in the same location as the original file. Is this new behavior intended?

And as it was the case in the other issue, this file fails to get deleted by BioFormats after considering it "invalid".

This error message showed up in a script that I ran after reopening a project I created in QuPath 0.3.0-SNAPSHOT

```
WARN: file deletion failed C:\Users\oburri\AppData\Local\Temp\qupath-memo-15327719930616011555\biop\Image_Processing\User_Projects\Olivier\USERNAME_REDACTED\Sample images - metadata\.MCF7_hPR_OVX_3rd_M6-10_Ki67.vsi.bfmemo
ERROR: temp file rename returned false: C:\Users\oburri\AppData\Local\Temp\qupath-memo-15327719930616011555\biop\Image_Processing\User_Projects\Olivier\valentina.scabia_UPBRI\Sample images - metadata\.MCF7_hPR_OVX_3rd_M6-10_Ki67.vsi.bfmemo5671337359207729134
ERROR: deleting invalid memo file: C:\Users\oburri\AppData\Local\Temp\qupath-memo-15327719930616011555\biop\Image_Processing\User_Projects\Olivier\valentina.scabia_UPBRI\Sample images - metadata\.MCF7_hPR_OVX_3rd_M6-10_Ki67.vsi.bfmemo
    at loci.formats.in.JPEGReader$DefaultJPEGReaderConstructorAccess.newInstance(Unknown Source)
    at com.esotericsoftware.kryo.Kryo$DefaultInstantiatorStrategy$1.newInstance(Kryo.java:1275)
    at com.esotericsoftware.kryo.Kryo.newInstance(Kryo.java:1139)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.create(FieldSerializer.java:562)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:538)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.read(DefaultArraySerializers.java:391)
    at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.read(DefaultArraySerializers.java:302)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.read(DefaultArraySerializers.java:391)
    at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.read(DefaultArraySerializers.java:302)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:709)
    at loci.formats.Memoizer$KryoDeser.loadReader(Memoizer.java:163)
    at loci.formats.Memoizer.loadMemo(Memoizer.java:888)
    at loci.formats.Memoizer.setId(Memoizer.java:666)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer$BioFormatsReaderManager.createReader(BioFormatsImageServer.java:1360)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer$BioFormatsReaderManager.createReader(BioFormatsImageServer.java:1265)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer$BioFormatsReaderManager.getReaderForThread(BioFormatsImageServer.java:1191)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer.getReader(BioFormatsImageServer.java:815)
    at qupath.lib.images.servers.bioformats.BioFormatsImageServer.readTile(BioFormatsImageServer.java:848)
    at qupath.lib.images.servers.AbstractTileableImageServer.getTile(AbstractTileableImageServer.java:184)
    at qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(AbstractTileableImageServer.java:275)
    at qupath.lib.images.servers.AbstractTileableImageServer.readBufferedImage(AbstractTileableImageServer.java:56)
    at qupath.lib.images.servers.ServerTools.getPaddedRequest(ServerTools.java:231)
    at qupath.opencv.ops.ImageOps$DefaultImageDataOp.apply(ImageOps.java:256)
    at qupath.tensorflow.stardist.StarDist2D.detectObjectsForTile(StarDist2D.java:807)
    at qupath.tensorflow.stardist.StarDist2D.lambda$detectObjects$5(StarDist2D.java:687)
    at java.base/java.util.stream.ReferencePipeline$7$1.accept(Unknown Source)
    at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
    at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
    at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
    at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
    at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
    at java.base/java.util.stream.AbstractTask.compute(Unknown Source)
    at java.base/java.util.concurrent.CountedCompleter.exec(Unknown Source)
    at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
    at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
    at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
    at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
    at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
WARN: file deletion failed C:\Users\oburri\AppData\Local\Temp\qupath-memo-15327719930616011555\biop\Image_Processing\User_Projects\Olivier\USERNAME_REDACTED\Sample images - metadata\.MCF7_hPR_OVX_3rd_M6-10_Ki67.vsi.bfmemo
```

**To Reproduce**
1. On Windows, create a new Project and add an Image that will use BioFormats
2. Display the image in the Viewer at least once
3. Close QuPath and restart
4. Reopen the file
5. Start a script and run it

**Desktop (please complete the following information):**
 - OS: Windows 10 build 20H2
 - QuPath Version 0.3.0 SNAPSHOT

Best regards,

Olivier

**Comment by petebankhead** [comment]
I don't recall any memoization-related changes in v0.3.0-SNAPSHOT, except for the fact it will be using a different JDK and updated version of Bio-Formats itself. Unless I'm missing a key part, the git blame suggests the memoization code was last changed 2 years ago.

v0.2 made a change from v0.1.2 to no longer store the .bfmemo files in the same directory as the image, because this was causing problems for some users and confusion with new files being generated unexpectedly.

Since v0.2, you can specify a memoization directory in the preferences, and also a memoization time in ms. If the time is <= 0, then no memoization files should be created. Otherwise, a file will be [generated either in the specified directory or in a temp directory](https://github.com/qupath/qupath/blob/dev-0.3/qupath-extension-bioformats/src/main/java/qupath/lib/images/servers/bioformats/BioFormatsImageServer.java#L1314).

Could it be that you have a directory set in the preferences? If you don't, I would expect the memoization files to be temporary and automatically removed.

In any case, the warning you see has been [logged by Bio-Formats](https://github.com/ome/bioformats/blob/25645389e076a7bd0011e04c4dd8982c0f0614ed/components/formats-bsd/src/loci/formats/Memoizer.java#L923). I don't know why it is unable to delete the file, but my guess is that it relates in some way to https://github.com/ome/bioformats/issues/3659 and the fact QuPath is now using Java 16. This is likely to cause various memoization problems involving kyro.

I can think of two potential workarounds:
* Add `--illegal-access=warn` to the Java options for QuPath (based on [this](https://openjdk.java.net/jeps/396))
* Turn off memoization entirely, at least by default

I'm not sure which is preferable; neither feels great.

I think you should be able to try the first option by editing the `.cfg` file - if you try, please let me know if it works

**Comment by petebankhead** [comment]
Sorry, I see now that the `AppData` folder must be the default choice for creating the temp file - so I presume you don't have a directory set.

I'm a bit confused that it's still picked up after restarting QuPath.

**Comment by petebankhead** [comment]
In the `QuPath.cfg` file of your installation, under `[JavaOptions]`, the exact line to add is
```
java-options=--illegal-access=warn
```
I've just tried this and it has resolved another problem that I expect to be related (the inability to add metadata to project entries in the current v0.3.0-SNAPSHOT).

**Comment by petebankhead** [comment]
@lacan I'll close this because the proposed fix has now been merged into `dev-0.3`. Please reopen if you try it and find the problem persists.
